duckduckgo/Android
# Bookmark Folders

Task/Issue URL: https://app.asana.com/0/72649045549333/1161200816074835/f
Tech Design URL: https://app.asana.com/0/481882893211075/1200496630620092/f
CC: 

**Description**:
This PR adds bookmark folder functionality.

**Steps to test this PR**:

- Freshly install the app.

> Add a folder

1. Go to "Bookmarks"
2. Tap the "add folder" action button to add a folder.
3. Verify that the "checkmark" action button isn't shown if there is no title.
4. Add a title.
5. Verify that the "checkmark" action button is shown.
6. Tap the "checkmark" action button.
7. Verify that a folder is created.
8. Verify that the folder's subtitle says "Empty".

> Add a subfolder

1. Tap on the newly created folder.
2. Tap the "add folder" action button and create subfolder/s.
3. Verify that the item count subtitle on the parent folder is correct.

> Edit a folder:

1. Edit one of the folders.
2. Verify that the "checkmark" action button isn't shown.
3. Delete the title.
4. Verify that the "checkmark" action button is not visible.
5. Push hardware back or the up arrow.
6. Verify that the folder title is not updated.
7. Edit the folder again.
8. Change the "location" of the folder.
9. Verify that the selected folder in the folder tree is correct.
10. Verify that the folder cannot be added to any of its subfolders.
11. Verify that the "checkmark" action button is shown.
12. Change the folder's title.
13. Tap the "checkmark" action button.
14. Verify that the folder's location and title have changed.

> Edit a bookmark (Browser):

1. Add a bookmark.
2. Click "Edit" on the SnackBar that is shown.
3. Verify that the "checkmark" action button is not shown.
4. Change the title, URL and location fields.
5. Verify that the "checkmark" action button behaves as expected.
6. Tap the "checkmark" action button.
7. Verify that the changes to the bookmark's title, URL and location are reflected.

> Edit a bookmark (Bookmarks list):

1. Add a bookmark.
2. Go to "Bookmarks".
3. Edit a bookmark in the list.
4. Verify that that the behavior is as expected.

> Edit a favorite (Browser):

1. Add a favorite.
2. Click "Edit" on the SnackBar that is shown.
3. Verify that the "checkmark" action button is not shown.
4. Change the title and URL.
5. Verify that the "checkmark" action button behaves as expected.
6. Tap the "checkmark" action button.
7. Verify that the changes to the favorite's title and URL are reflected.

> Edit a favorite (Favorites list):

1. Add a favorite.
2. Go to "Bookmarks".
3. Edit a favorite in the list.
4. Verify that that the behavior is as expected.

> Deletion (Empty folder):

1. Delete an empty folder.
2. Verify that the folder is immediately deleted.
3. Tap the "Undo" action in the SnackBar that is shown.
4. Verify that the folder re-added.

> Deletion (Non-empty folder):

1. Delete a folder with subfolders and bookmarks.
2. Verify that a dialog is shown with the correct folder title and item count.
3. Tap "Cancel".
4. Verify that the folder is not deleted.
5. Delete the folder again and tap "Delete" on the dialog.
6. Verify that the folder is deleted.
7. Tap the "Undo" action in the SnackBar that is shown.
8. Verify that the folder (And its contents) are re-added.

> Search:

1. Verify that the "search" action button is not shown if there are no bookmarks or folders.
2. Add some bookmarks and folders.
3. Verify that the "search" action button is shown.
4. Tap the "search" action button.
5. Verify that all of the bookmarks and folders are shown in the list.
6. Verify that the folders have the correct item counts.
7. Tap a folder.
8. Verify that the folder is navigated to successfully.
9. Verify that the opened folder has the "search" action button even if it is empty.
10. Tap the search button.
11. Tap a bookmark.
12. Verify that the bookmark opens successfully.

> Import/export:

1. Add some bookmarks and create some folders (Make sure to add some subfolders with bookmarks).
2. Export the bookmarks using the overflow menu.
3. Check that the confirmation SnackBar is shown.
4. Delete all bookmarks and folders.
4. Import the bookmarks back into DuckDuckGo.
5. Verify that the bookmarks and folder structure were imported correctly.
6. Importing the same file twice, should import the same folders twice.

> Import/export from external:

1. Open the folder at this location: [app/src/androidTest/resources/bookmarks/](https://github.com/duckduckgo/Android/tree/feature/josh/bookmark-folders/app/src/androidTest/resources/bookmarks).
2. Copy these files onto a test device/emulator.
3. Import each of these files into DuckDuckGo.
4. Verify that these files import correctly into DuckDuckGo except for [bookmarks_invalid.html](https://github.com/duckduckgo/Android/blob/feature/josh/bookmark-folders/app/src/androidTest/resources/bookmarks/bookmarks_invalid.html).

> Test migration:

1. Switch to develop and install the app.
2. Switch to [feature/josh/bookmark-folders](https://github.com/duckduckgo/Android/tree/feature/josh/bookmark-folders) and install the app.
3. Verify that the app upgrades successfully and opens as expected.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
